From a3241ca7a418f2c8762b808cc12bf1e7c7ac2d60 Mon Sep 17 00:00:00 2001
From: Colm O Heigeartaigh <coheigea@apache.org>
Date: Fri, 22 Jul 2011 14:48:44 +0000
Subject: [PATCH] [SANTUARIO-284] - ElementProxy#getTextFromChildElement()
 doesn't get all of the text if the element contains an entity like &amp;  -
 Fix applied on trunk, thanks!

git-svn-id: https://svn.apache.org/repos/asf/santuario/xml-security-java/trunk@1149625 13f79535-47bb-0310-9956-ffa450edef68
---
 CHANGELOG.txt                                             | 1 +
 .../java/org/apache/xml/security/utils/ElementProxy.java  | 8 ++------
 2 files changed, 3 insertions(+), 6 deletions(-)

diff --git a/CHANGELOG.txt b/CHANGELOG.txt
index b5ca997bc9..7841e440d2 100644
--- a/CHANGELOG.txt
+++ b/CHANGELOG.txt
@@ -1,6 +1,7 @@
 Changelog for "Apache xml-security" <http://santuario.apache.org/>
 
 New in v1.5.0-SNAPSHOT
+    Fixed SANTUARIO-284: ElementProxy#getTextFromChildElement() doesn't get all of the text if the element contains an entity like &amp;
     Fixed SANTUARIO-273: xml:base attribute not processed correctly in C14N11 canonicalization.
     Fixed SANTUARIO-260: Review logging
     Fixed SANTUARIO-257: Use JUnit4 for testing
diff --git a/src/main/java/org/apache/xml/security/utils/ElementProxy.java b/src/main/java/org/apache/xml/security/utils/ElementProxy.java
index 68711e255d..8429667954 100644
--- a/src/main/java/org/apache/xml/security/utils/ElementProxy.java
+++ b/src/main/java/org/apache/xml/security/utils/ElementProxy.java
@@ -370,15 +370,11 @@ public byte[] getBytesFromChildElement(String localname, String namespace)
      * @return the Text of the textNode
      */
     public String getTextFromChildElement(String localname, String namespace) {
-        Text t =
-            (Text) XMLUtils.selectNode(
+        return XMLUtils.selectNode(
                 this.constructionElement.getFirstChild(),
                 namespace,
                 localname,
-                0
-            ).getFirstChild();
-
-        return t.getData();      
+                0).getTextContent();
     }
 
     /**
